virtual-desktop/src/app/authentication-manager/login/login.component.ts

```diff
@@ -324,50 +297,12 @@ export class LoginComponent implements OnInit {
       error => {
         this.needLogin = true;
         let jsonMessage = error.json();
-        if (jsonMessage) {
-          if (jsonMessage.categories) {
-            let failedTypes = [];
-            let failedPlugins = new Set<string>();
-            let keys = Object.keys(jsonMessage.categories);
-            for (let i = 0; i < keys.length; i++) {
-              if (!jsonMessage.categories[keys[i]].success) {
-                failedTypes.push(keys[i]);
-              }
-              let plugins = Object.keys(jsonMessage.categories[keys[i]].plugins);
-              for (let j = 0; j < plugins.length; j++) {
-                if (jsonMessage.categories[keys[i]].plugins[plugins[j]].canChangePassword) {
-                  this.passwordServices.add(plugins[j]);
-                }
-              }
-            }
-            if (error.status == HTTP_STATUS_PRECONDITION_REQUIRED) {
-              this.expiredPassword = true;
-              this.loginMessage = this.translation.translate(PASSWORD_EXPIRED);
-            } else {
-              this.errorDetails = '';
-              let err;
-              // Get the server error messages from auth plugins
-              for (let i = 0; i < failedTypes.length; i++) {
-                let plugins = Object.keys(jsonMessage.categories[failedTypes[i]].plugins);
-                for (let j = 0; j < plugins.length; j++) {
-                  err = jsonMessage.categories[failedTypes[i]].plugins[plugins[j]].error;
-                  if(err) {
-                    this.errorMessage = err.message;
-                    if(!failedPlugins.has(plugins[j])) {
-                      // Appending the error messages and corresponding unique plugin-ids that have errors
-                      this.errorDetails += `${plugins[j]}: ${this.errorMessage}.\n${err.body}\n`;
-                      failedPlugins.add(plugins[j])
-                    }
-                  }
-                }
-              }
-              if(!err) {
-                this.errorMessage = this.translation.translate('AuthenticationFailed',
-                { numTypes: failedTypes.length, types: JSON.stringify(failedTypes) });
-              }
-              this.password = '';
-            }
-          }
+        if (error.status == HTTP_STATUS_PRECONDITION_REQUIRED) {
+          this.expiredPassword = true;
+          this.loginMessage = this.translation.translate(PASSWORD_EXPIRED);
+        } else if(jsonMessage) {
```

Comment:
I think this breaks the expired password reset ability.
Basically, we need to know what plugins can handle password reset before we can let the user do a password reset. The answer is in the jsonMessage, so you can't if-else this. You need to keep the PASSWORD_EXPIRED check logic within the jsonMessage handling. This might mean that you can't de-duplicate the displayErrorDetails function without knowing which block of code is calling it since there are 2: `attemptLogin` and `ngOnInit`